### PR TITLE
Dbn inference

### PR DESCRIPTION
**Pull request by palashahuja**

### CODE REVIEW

**Comment by palashahuja** [comment]
@ankurankan ready for a review.

File: pgmpy/inference/__init__.py [hunk 1]
```diff
@@ -2,10 +2,10 @@
 from .ExactInference import VariableElimination
 from .ExactInference import BeliefPropagation
 from .Sampling import BayesianModelSampling
-from .mplp import Mplp
```
Comment (ankurankan) [inline pgmpy/inference/__init__.py:5]: @palashahuja Why have you removed Mplp?

File: pgmpy/inference/__init__.py [hunk 2]
```diff
@@ -2,10 +2,10 @@
...
+from .dbn_inference import DBNInference
 
 __all__ = ['Inference',
            'VariableElimination',
+           'DBNInference',
            'BeliefPropagation',
-           'BayesianModelSampling',
-           'Mplp']
```
Comment (ankurankan) [inline pgmpy/inference/__init__.py:11]: Same here

File: pgmpy/inference/dbn_inference.py [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+from pgmpy.inference import Inference, BeliefPropagation
+
+
+class DBNInference(Inference):
+
+    def __init__(self, model):
+        """
+        Class for performing inference using Belief Propagation method
+        for the input Dynamic Bayesian Network.
+        Parameters
```
Comment (ankurankan) [inline pgmpy/inference/dbn_inference.py:17]: @palashahuja Please check the docstrings written in other methods and follow the same format. Also add doctests.

File: pgmpy/inference/dbn_inference.py [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+        model for which inference is to performed
+        """
+        super().__init__(model)
+        self.interface_nodes_0 = model.get_interface_nodes(0)
+        self.interface_nodes_1 = model.get_interface_nodes(1)
+
+        start_markov_model = self.start_bayesian_model.to_markov_model()
+        one_and_half_markov_model = self.one_and_half_model.to_markov_model()
+
+        combinations_slice_0 = list(itertools.combinations(model.get_interface_nodes(0), 2))
```
Comment (ankurankan) [inline pgmpy/inference/dbn_inference.py]: @palashahuja No need to convert these to lists.

File: pgmpy/inference/dbn_inference.py [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+        combinations_slice_1 = list(itertools.combinations(model.get_interface_nodes(1), 2))
+
+        start_markov_model.add_edges_from(combinations_slice_0)
```
Comment (ankurankan) [inline pgmpy/inference/dbn_inference.py]: `add_edges_from` can accept any iterable so can directly pass the combinations object.

File: pgmpy/inference/dbn_inference.py [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+        one_and_half_markov_model.add_edges_from(combinations_slice_0 + combinations_slice_1)
```
Comment (ankurankan) [inline pgmpy/inference/dbn_inference.py]: Can use `itertools.chain` here.

File: pgmpy/inference/dbn_inference.py [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+        """
+        marginalizing the factor selectively for a set of variables.
+        Parameters
+        ----------
+        nodes: iterable container
+            A container of nodes (list, dict, set, etc.).
+        factor: factor
+            factor which is to be marginalized.
+        """
+        marginalizing_nodes = list(set(factor.scope()).difference(nodes))
```
Comment (ankurankan) [inline pgmpy/inference/dbn_inference.py:137]: @palashahuja No need to convert it to list. `marginalize` should work with sets also

File: pgmpy/inference/dbn_inference.py [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+        belief_prop: Belief Propagation
+            Belief Propagation which needs to be updated.
+        in_clique: clique
+            The factor which needs to be updated corresponding to the input clique.
+        out_clique_potential: factor
+            Multiplying factor which will be multiplied to the factor corresponding to the clique.
+        """
+        old_factor = belief_prop.junction_tree.get_factors(in_clique)
+        belief_prop.junction_tree.remove_factors(old_factor)
+        new_factor = old_factor*out_clique_potential
```
Comment (ankurankan) [inline pgmpy/inference/dbn_inference.py]: PEP 8

Comment (palashahuja) [inline pgmpy/inference/dbn_inference.py]: This doesn't seem to be a PEP 8 error. I have already checked for them .

File: pgmpy/inference/dbn_inference.py [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+        start_bp.calibrate()
+        mid_bp.calibrate()
+        evidence_0 = self._get_evidence(evidence, 0, 0)
+        interface_nodes_dict = {}
+        if evidence is not None:
+            interface_nodes_dict = {k: v for k, v in evidence_0.items() if k in self.interface_nodes_0}
+        initial_factor = self._get_factor(start_bp, evidence_0)
+        marginalized_factor = self._marginalize_factor(self.interface_nodes_0, initial_factor)
+        self._update_belief(mid_bp, self.in_clique, marginalized_factor)
+        if variable_dict[0] != []:
```
Comment (ankurankan) [inline pgmpy/inference/dbn_inference.py]: Can't simply do `if variable_dict[0]:` ??

File: pgmpy/inference/dbn_inference.py [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+        for var in variables:
+                variable_dict[var[1]].append(var)
+        time_range = max(variable_dict)
+        start_bp = BeliefPropagation(self.start_junction_tree)
+        mid_bp = BeliefPropagation(self.one_and_half_junction_tree)
+        start_bp.calibrate()
+        mid_bp.calibrate()
+        evidence_0 = self._get_evidence(evidence, 0, 0)
+        interface_nodes_dict = {}
+        if evidence is not None:
```
Comment (ankurankan) [inline pgmpy/inference/dbn_inference.py]: Can't simply do `if evidence:`?

File: pgmpy/inference/dbn_inference.py [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+        marginalized_factor = self._marginalize_factor(self.interface_nodes_0, initial_factor)
+        self._update_belief(mid_bp, self.in_clique, marginalized_factor)
+        if variable_dict[0] != []:
+            factor_values = start_bp.query(variable_dict[0], evidence=evidence_0)
+        else:
+            factor_values = {}
+
+        for time in range(1, time_range + 1):
+            evidence_time = self._get_evidence(evidence, time, 1)
+            if interface_nodes_dict != {}:
```
Comment (ankurankan) [inline pgmpy/inference/dbn_inference.py]: Same here

File: pgmpy/inference/dbn_inference.py [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+                if evidence_time is not None:
+                    evidence_time.update(interface_nodes_dict)
+                else:
+                    evidence_time = interface_nodes_dict
+
+            if variable_dict[time] != []:
```
Comment (ankurankan) [inline pgmpy/inference/dbn_inference.py]: Same here

File: pgmpy/inference/dbn_inference.py (hunk 7 above)
Comment (palashahuja) [inline pgmpy/inference/dbn_inference.py:137]: @ankurankan `marginalize` does not work with sets.

```
>>> phi = Factor(['a','b'],[2,2],range(4))
>>> phi.marginalize(set('b'))
>>> Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/palash/pgmpy/pgmpy/factors/Factor.py", line 195, in marginalize
    if variable not in self.variables:
TypeError: unhashable type: 'set'
```

File: pgmpy/inference/dbn_inference.py [hunk 13]
```diff
@@ -0,0 +1,238 @@
...
+                                               [0.4, 0.25, 0.08, 0.3],
+                                               [0.3, 0.7, 0.2, 0.2]],[('D',0),('I',0)],[2,2])
+        >>> d_i_cpd = TabularCPD(('D',1),2,[[0.6,0.3],[0.4,0.7]],[('D',0)],2)
+        >>> diff_cpd = TabularCPD(('D',0),2,[[0.6,0.4]])
+        >>> intel_cpd = TabularCPD(('I',0),2,[[0.7,0.3]])
+        >>> i_i_cpd = TabularCPD(('I',1),2,[[0.5,0.4],[0.5,0.6]],[('I',0)],2)
+        >>> dbnet.add_edges_from([(('D', 0), ('D', 1)), (('I', 0), ('I', 1)), (('D',1),('G',1)), (('I',1), ('G',1))])
+        >>> dbnet.add_cpds(grade_cpd, d_i_cpd, diff_cpd, intel_cpd, i_i_cpd)
+        >>> dbnet.initialize_initial_state()
+        >>> dbn_inf = DBNInference(dbnet)
```
Comment (ankurankan) [inline pgmpy/inference/dbn_inference.py:240]: @palashahuja Getting this error while trying to run this: 

``` python
In [15]: dbn_inf = DBNInference(dbnet)
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-15-545100381a33> in <module>()
----> 1 dbn_inf = DBNInference(dbnet)

/home/ubuntu/pgmpy/pgmpy/inference/dbn_inference.py in __init__(self, model)
     61 
     62         self.start_interface_clique = self._get_clique(self.start_junction_tree, self.interface_nodes_0)
---> 63         self.in_clique = self._get_clique(self.one_and_half_junction_tree, self.interface_nodes_0)
     64         self.out_clique = self._get_clique(self.one_and_half_junction_tree, self.interface_nodes_1)
     65 

/home/ubuntu/pgmpy/pgmpy/inference/dbn_inference.py in _get_clique(self, junction_tree, nodes)
     87             A container of nodes (list, dict, set, etc.).
     88         """
---> 89         return [clique for clique in junction_tree.nodes() if set(nodes).issubset(clique)][0]
     90 
     91     def _get_evidence(self, evidence_dict, time, shift):

IndexError: list index out of range

```

**Comment by palashahuja** [comment]
@ankurankan  Code ready for review

**Pull request closed without merging**